Readme.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> MQTT Messaging for Golang

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The MQTT module contains a set of components for messaging using the Mqtt protocol. Contains the implementation of the components for working with messages: MqttMessageQueue, MqttConnectionResolver.

The module contains the following packages:
- **Build** - factory default implementation
- **Connect** - components for setting up the connection to the MQTT broker
- **Queues** - components of working with a message queue via the MQTT protocol

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/concepts/configuration/)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-mqtt-gox/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)


## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-mqtt-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The Golang version of Pip.Services is created and maintained by **Sergey Seroukhov** and **Levichev Dmitry**.

The documentation is written by:
- **Levichev Dmitry**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> MQTT Messaging for Golang Changelog

## <a name="1.0.1"></a> 1.0.1 (2022-10-10)

- Updated documentation and dependencies

## <a name="1.0.0"></a> 1.0.0 (2022-07-10)

Initial public release

### Features
* **build** factory default implementation
* **connect** components for setting up the connection to the MQTT broker
* **queues** components of working with a message queue

test/queues/MessageQueueFixture.go
package test_queues

import (
	"context"
	"sync"
	"testing"
	"time"

	"github.com/pip-services3-gox/pip-services3-messaging-gox/queues"
	"github.com/stretchr/testify/assert"
)

type MessageQueueFixture struct {
	queue queues.IMessageQueue
}

func NewMessageQueueFixture(queue queues.IMessageQueue) *MessageQueueFixture {
	c := MessageQueueFixture{
		queue: queue,
	}
	return &c
}

func (c *MessageQueueFixture) TestSendReceiveMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// if c.queue.GetCapabilities().CanMessageCount() {
	// 	count, rdErr := c.queue.MessageCount()
	// 	assert.Nil(t, rdErr)
	// 	assert.Greater(t, count, (int64)(0))
	// }

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestMessageCount(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	c.queue.Send(context.Background(), "", envelope1)

	time.Sleep(500 * time.Millisecond)

	count, err := c.queue.ReadMessageCount()
	assert.Nil(t, err)
	assert.True(t, count >= 1)
}

func (c *MessageQueueFixture) TestReceiveSendMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))

	time.AfterFunc(500*time.Millisecond, func() {
		c.queue.Send(context.Background(), "", envelope1)
	})

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestReceiveCompleteMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	// count, rdErr := c.queue.ReadMessageCount()
	// assert.Nil(t, rdErr)
	// assert.Greater(t, count, (int64)(0))

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	cplErr := c.queue.Complete(context.Background(), envelope2)
	assert.Nil(t, cplErr)
	assert.Nil(t, envelope2.GetReference())
}

func (c *MessageQueueFixture) TestReceiveAbandonMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	abdErr := c.queue.Abandon(context.Background(), envelope2)
	assert.Nil(t, abdErr)

	envelope2, rcvErr = c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)
}

func (c *MessageQueueFixture) TestSendPeekMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(100 * time.Millisecond)

	envelope2, pkErr := c.queue.Peek(context.Background(), "")
	assert.Nil(t, pkErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	// pop message from queue for next test
	_, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
}

func (c *MessageQueueFixture) TestPeekNoMessage(t *testing.T) {
	envelope, pkErr := c.queue.Peek(context.Background(), "")
	assert.Nil(t, pkErr)
	assert.Nil(t, envelope)
}

func (c *MessageQueueFixture) TestMoveToDeadMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	envelope2, rcvErr := c.queue.Receive(context.Background(), "", 10000*time.Millisecond)
	assert.Nil(t, rcvErr)
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	mvErr := c.queue.MoveToDeadLetter(context.Background(), envelope2)
	assert.Nil(t, mvErr)
}

func (c *MessageQueueFixture) TestOnMessage(t *testing.T) {
	envelope1 := queues.NewMessageEnvelope("123", "Test", []byte("Test message"))
	receiver := &TestMsgReceiver{}
	c.queue.BeginListen(context.Background(), "", receiver)

	time.Sleep(1000 * time.Millisecond)

	sndErr := c.queue.Send(context.Background(), "", envelope1)
	assert.Nil(t, sndErr)

	time.Sleep(1000 * time.Millisecond)

	envelope2 := receiver.GetEnvelope()
	assert.NotNil(t, envelope2)
	assert.Equal(t, envelope1.MessageType, envelope2.MessageType)
	assert.Equal(t, envelope1.Message, envelope2.Message)
	assert.Equal(t, envelope1.CorrelationId, envelope2.CorrelationId)

	c.queue.EndListen(context.Background(), "")
}

type TestMsgReceiver struct {
	_envelope *queues.MessageEnvelope
	_lock     sync.Mutex
}

func (c *TestMsgReceiver) GetEnvelope() queues.MessageEnvelope {
	c._lock.Lock()
	defer c._lock.Unlock()
	return *c._envelope
}

func (c *TestMsgReceiver) ReceiveMessage(ctx context.Context, envelope *queues.MessageEnvelope, queue queues.IMessageQueue) (err error) {
	c._lock.Lock()
	c._envelope = envelope
	c._lock.Unlock()
	return nil
}


test/connect/MqttConnection_test.go
package test_connect

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	connect "github.com/pip-services3-gox/pip-services3-mqtt-gox/connect"
	"github.com/stretchr/testify/assert"
)

type mqttConnectionTest struct {
	connection *connect.MqttConnection
}

func newMqttConnectionTest() *mqttConnectionTest {
	mqttUri := os.Getenv("MQTT_SERVICE_URI")
	mqttHost := os.Getenv("MQTT_SERVICE_HOST")
	if mqttHost == "" {
		mqttHost = "localhost"
	}

	mqttPort := os.Getenv("MQTT_SERVICE_PORT")
	if mqttPort == "" {
		mqttPort = "1883"
	}

	mqttUser := os.Getenv("MQTT_USER")
	// if mqttUser == "" {
	// 	mqttUser = ""
	// }
	mqttPassword := os.Getenv("MQTT_PASS")
	// if mqttPassword == "" {
	// 	mqttPassword = ""
	// }

	if mqttUri == "" && mqttHost == "" {
		return nil
	}

	connection := connect.NewMqttConnection()
	connection.Configure(context.Background(), cconf.NewConfigParamsFromTuples(
		"connection.uri", mqttUri,
		"connection.host", mqttHost,
		"connection.port", mqttPort,
		"credential.username", mqttUser,
		"credential.password", mqttPassword,
	))

	return &mqttConnectionTest{
		connection: connection,
	}
}

func (c *mqttConnectionTest) TestOpenClose(t *testing.T) {
	err := c.connection.Open(context.Background(), "")
	assert.Nil(t, err)
	assert.True(t, c.connection.IsOpen())
	assert.NotNil(t, c.connection.GetConnection())

	err = c.connection.Close(context.Background(), "")
	assert.Nil(t, err)
	assert.False(t, c.connection.IsOpen())
	assert.Nil(t, c.connection.GetConnection())
}

func TestMqttConnection(t *testing.T) {
	c := newMqttConnectionTest()
	if c == nil {
		return
	}

	t.Run("Open and Close", c.TestOpenClose)
}


queues/MqttMessageQueue.go
package queues

import (
	"context"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	clog "github.com/pip-services3-gox/pip-services3-components-gox/log"
	cqueues "github.com/pip-services3-gox/pip-services3-messaging-gox/queues"
	connect "github.com/pip-services3-gox/pip-services3-mqtt-gox/connect"
)

//
// MqttMessageQueue are message queue that sends and receives messages via MQTT message broker.
//
// Configuration parameters:
//
//	- topic:                         name of MQTT topic to subscribe
//	- connection(s):
//		- discovery_key:               (optional) a key to retrieve the connection from  IDiscovery
//		- host:                        host name or IP address
//		- port:                        port number
//		- uri:                         resource URI or connection string with all parameters in it
//	- credential(s):
//		- store_key:                   (optional) a key to retrieve the credentials from  ICredentialStore
//		- username:                    user name
//		- password:                    user password
//	- options:
//		- serialize_envelope:    (optional) true to serialize entire message as JSON, false to send only message payload (default: true)
//		- autosubscribe:        (optional) true to automatically subscribe on option (default: false)
//		- qos:                  (optional) quality of service level aka QOS (default: 0)
//		- retain:               (optional) retention flag for published messages (default: false)
//		- retry_connect:        (optional) turns on/off automated reconnect when connection is log (default: true)
//		- connect_timeout:      (optional) number of milliseconds to wait for connection (default: 30000)
//		- reconnect_timeout:    (optional) number of milliseconds to wait on each reconnection attempt (default: 1000)
//		- keepalive_timeout:    (optional) number of milliseconds to ping broker while inactive (default: 3000)
//
//
// References:
//	- *:logger:*:*:1.0             (optional)  ILogger components to pass log messages
//	- *:counters:*:*:1.0           (optional)  ICounters components to pass collected measurements
//	- *:discovery:*:*:1.0          (optional)  IDiscovery services to resolve connections
//	- *:credential-store:*:*:1.0   (optional) Credential stores to resolve credentials
//	- *:connection:mqtt:*:1.0      (optional) Shared connection to MQTT service
//
// See MessageQueue
// See MessagingCapabilities
//
// Example:
//
//	queue := NewMqttMessageQueue("myqueue")
//	queue.Configure(context.Background(), cconf.NewConfigParamsFromTuples(
//		"subject", "mytopic",
//		"connection.protocol", "mqtt"
//		"connection.host", "localhost"
//		"connection.port", 1883
//	))
//
//	queue.open(context.Background(), "123")
//
//	queue.Send(context.Background(), "123", NewMessageEnvelope("", "mymessage", "ABC"))
//
//	message, err := queue.Receive(context.Background(), "123")
//	if (message != nil) {
//		...
//		queue.Complete(context.Background(), "123", message);
//	}
type MqttMessageQueue struct {
	*cqueues.MessageQueue

	defaultConfig   *cconf.ConfigParams
	config          *cconf.ConfigParams
	references      cref.IReferences
	opened          bool
	localConnection bool

	//The dependency resolver.
	DependencyResolver *cref.DependencyResolver
	//The logger.
	Logger *clog.CompositeLogger
	//The MQTT connection component.
	Connection *connect.MqttConnection

	serializeEnvelope bool
	topic             string
	qos               byte
	retain            bool
	autoSubscribe     bool
	subscribed        bool
	messages          []cqueues.MessageEnvelope
	receiver          cqueues.IMessageReceiver
}

// Creates a new instance of the queue component.
//   - name    (optional) a queue name.
func NewMqttMessageQueue(name string) *MqttMessageQueue {
	c := MqttMessageQueue{
		defaultConfig: cconf.NewConfigParamsFromTuples(
			"topic", nil,
			"options.autosubscribe", false,
			"options.serialize_envelope", false,
			"options.retry_connect", true,
			"options.connect_timeout", 30000,
			"options.reconnect_timeout", 1000,
			"options.keepalive_timeout", 60000,
			"options.qos", 0,
			"options.retain", false,
		),
		Logger: clog.NewCompositeLogger(),
	}
	c.MessageQueue = cqueues.InheritMessageQueue(&c, name,
		cqueues.NewMessagingCapabilities(false, true, true, true, true, false, false, false, true))
	c.DependencyResolver = cref.NewDependencyResolver()
	c.DependencyResolver.Put(context.Background(), "connection", cref.NewDescriptor("pip-services", "connection", "mqtt", "*", "1.0"))
	c.DependencyResolver.Configure(context.Background(), c.defaultConfig)

	c.messages = make([]cqueues.MessageEnvelope, 0)

	return &c
}

// Configures component by passing configuration parameters.
// Parameters:
//	- ctx context.Context	operation context.
//	- config    configuration parameters to be set.
func (c *MqttMessageQueue) Configure(ctx context.Context, config *cconf.ConfigParams) {
	config = config.SetDefaults(c.defaultConfig)
	c.config = config

	c.DependencyResolver.Configure(ctx, config)

	c.topic = config.GetAsStringWithDefault("topic", c.topic)
	c.serializeEnvelope = config.GetAsBooleanWithDefault("options.serialize_envelope", c.serializeEnvelope)
	c.autoSubscribe = config.GetAsBooleanWithDefault("options.autosubscribe", c.autoSubscribe)
	c.qos = byte(config.GetAsIntegerWithDefault("options.qos", int(c.qos)))
	c.retain = config.GetAsBooleanWithDefault("options.retain", c.retain)
}

// Sets references to dependent components.
// Parameters:
//	- ctx context.Context	operation context.
//	- references 	references to locate the component dependencies.
func (c *MqttMessageQueue) SetReferences(ctx context.Context, references cref.IReferences) {
	c.references = references
	c.Logger.SetReferences(ctx, references)

	// Get connection
	c.DependencyResolver.SetReferences(ctx, references)
	result := c.DependencyResolver.GetOneOptional("connection")
	if dep, ok := result.(*connect.MqttConnection); ok {
		c.Connection = dep
	}
	// Or create a local one
	if c.Connection == nil {
		c.Connection = c.createConnection()
		c.localConnection = true
	} else {
		c.localConnection = false
	}
}

// Unsets (clears) previously set references to dependent components.
func (c *MqttMessageQueue) UnsetReferences() {
	c.Connection = nil
}

func (c *MqttMessageQueue) createConnection() *connect.MqttConnection {
	connection := connect.NewMqttConnection()
	if c.config != nil {
		connection.Configure(context.Background(), c.config)
	}
	if c.references != nil {
		connection.SetReferences(context.Background(), c.references)
	}
	return connection
}

// Checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *MqttMessageQueue) IsOpen() bool {
	return c.opened
}

// Opens the component.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId 	(optional) transaction id to trace execution through call chain.
//	- Returns 			 error or nil no errors occured.
func (c *MqttMessageQueue) Open(ctx context.Context, correlationId string) (err error) {
	if c.opened {
		return nil
	}

	if c.Connection == nil {
		c.Connection = c.createConnection()
		c.localConnection = true
	}

	if c.localConnection {
		err = c.Connection.Open(ctx, correlationId)
	}

	if err == nil && c.Connection == nil {
		err = cerr.NewInvalidStateError(correlationId, "NO_CONNECTION", "MQTT connection is missing")
	}

	if err == nil && !c.Connection.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "CONNECT_FAILED", "MQTT connection is not opened")
	}

	if err != nil {
		return err
	}

	// Automatically subscribe if needed
	if c.autoSubscribe {
		err = c.subscribe(ctx, correlationId)
		if err != nil {
			return err
		}
	}

	c.opened = true

	return err
}

// Closes component and frees used resources.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId 	(optional) transaction id to trace execution through call chain.
//	- Returns 			error or nil no errors occured.
func (c *MqttMessageQueue) Close(ctx context.Context, correlationId string) (err error) {
	if !c.opened {
		return nil
	}

	if c.Connection == nil {
		return cerr.NewInvalidStateError(correlationId, "NO_CONNECTION", "MQTT connection is missing")
	}

	if c.localConnection {
		err = c.Connection.Close(ctx, correlationId)
	}
	if err != nil {
		return err
	}

	// Unsubscribe from topic
	if c.subscribed {
		topic := c.getTopic()
		c.Connection.Unsubscribe(ctx, topic, c)
		c.subscribed = false
	}

	c.Lock.Lock()
	defer c.Lock.Unlock()
	c.opened = false
	c.receiver = nil
	c.messages = make([]cqueues.MessageEnvelope, 0)

	return nil
}

func (c *MqttMessageQueue) getTopic() string {
	if c.topic != "" {
		return c.topic
	}
	return c.Name()
}

func (c *MqttMessageQueue) subscribe(ctx context.Context, correlationId string) error {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	// Check if already were subscribed
	if c.subscribed {
		return nil
	}

	// Subscribe to the topic
	topic := c.getTopic()
	err := c.Connection.Subscribe(ctx, topic, c.qos, c)
	if err != nil {
		c.Logger.Error(ctx, correlationId, err, "Failed to subscribe to topic "+topic)
		return err
	}

	c.subscribed = true
	return nil
}

func (c *MqttMessageQueue) fromMessage(message *cqueues.MessageEnvelope) ([]byte, error) {
	if message == nil {
		return nil, nil
	}

	data := message.Message
	if c.serializeEnvelope {
		message.SentTime = time.Now()
		var err error
		data, err = json.Marshal(message)
		if err != nil {
			return nil, err
		}
	}

	return data, nil
}

func (c *MqttMessageQueue) toMessage(msg mqtt.Message) (*cqueues.MessageEnvelope, error) {
	message := cqueues.NewEmptyMessageEnvelope()

	if c.serializeEnvelope {
		err := json.Unmarshal(msg.Payload(), message)
		if err != nil {
			return nil, err
		}
	} else {
		message.MessageId = strconv.FormatUint(uint64(msg.MessageID()), 10)
		message.MessageType = msg.Topic()
		message.Message = msg.Payload()
	}
	message.SetReference(msg)

	return message, nil
}

func (c *MqttMessageQueue) OnMessage(msg mqtt.Message) {
	// Skip if it came from a wrong topic
	expectedTopic := c.getTopic()
	if !strings.Contains(expectedTopic, "*") && expectedTopic != msg.Topic() {
		return
	}

	// Deserialize message
	message, err := c.toMessage(msg)
	if message == nil || err != nil {
		c.Logger.Error(context.Background(), "", err, "Failed to read received message")
		return
	}

	c.Counters.IncrementOne(context.Background(), "queue."+c.Name()+".received_messages")
	c.Logger.Debug(context.Background(), message.CorrelationId, "Received message %s via %s", message, c.Name())

	// Send message to receiver if its set or put it into the queue
	c.Lock.Lock()
	defer c.Lock.Unlock()
	if c.receiver != nil {
		receiver := c.receiver
		c.sendMessageToReceiver(context.Background(), receiver, message)
	} else {
		c.messages = append(c.messages, *message)
	}
}

// Clear method are clears component state.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId 	string (optional) transaction id to trace execution through call chain.
// Returns error or nil no errors occured.
func (c *MqttMessageQueue) Clear(ctx context.Context, correlationId string) error {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	c.messages = make([]cqueues.MessageEnvelope, 0)

	return nil
}

// ReadMessageCount method are reads the current number of messages in the queue to be delivered.
// Returns number of messages or error.
func (c *MqttMessageQueue) ReadMessageCount() (int64, error) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	count := (int64)(len(c.messages))
	return count, nil
}

// Peek method are peeks a single incoming message from the queue without removing it.
// If there are no messages available in the queue it returns nil.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId  string  (optional) transaction id to trace execution through call chain.
// Returns: result *cqueues.MessageEnvelope, err error
// message or error.
func (c *MqttMessageQueue) Peek(ctx context.Context, correlationId string) (*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return nil, err
	}

	var message *cqueues.MessageEnvelope

	// Pick a message
	c.Lock.Lock()
	if len(c.messages) > 0 {
		message = &c.messages[0]
	}
	c.Lock.Unlock()

	if message != nil {
		c.Logger.Trace(ctx, message.CorrelationId, "Peeked message %s on %s", message, c.String())
	}

	return message, nil
}

// PeekBatch method are peeks multiple incoming messages from the queue without removing them.
// If there are no messages available in the queue it returns an empty list.
// Important: This method is not supported by MQTT.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId     (optional) transaction id to trace execution through call chain.
//	- messageCount      a maximum number of messages to peek.
// Returns:          callback function that receives a list with messages or error.
func (c *MqttMessageQueue) PeekBatch(ctx context.Context, correlationId string, messageCount int64) ([]*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return nil, err
	}

	c.Lock.Lock()
	batchMessages := c.messages
	if messageCount <= (int64)(len(batchMessages)) {
		batchMessages = batchMessages[0:messageCount]
	}
	c.Lock.Unlock()

	messages := []*cqueues.MessageEnvelope{}
	for _, message := range batchMessages {
		messages = append(messages, &message)
	}

	c.Logger.Trace(ctx, correlationId, "Peeked %d messages on %s", len(messages), c.Name())

	return messages, nil
}

// Receive method are receives an incoming message and removes it from the queue.
// Parameters:
//	- ctx context.Context	operation context.
//  - correlationId   string   (optional) transaction id to trace execution through call chain.
//  - waitTimeout  time.Duration     a timeout in milliseconds to wait for a message to come.
// Returns:  result *cqueues.MessageEnvelope, err error
// receives a message or error.
func (c *MqttMessageQueue) Receive(ctx context.Context, correlationId string, waitTimeout time.Duration) (*cqueues.MessageEnvelope, error) {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return nil, err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return nil, err
	}

	messageReceived := false
	var message *cqueues.MessageEnvelope
	elapsedTime := time.Duration(0)

	for elapsedTime < waitTimeout && !messageReceived {
		c.Lock.Lock()
		if len(c.messages) == 0 {
			c.Lock.Unlock()
			time.Sleep(time.Duration(100) * time.Millisecond)
			elapsedTime += time.Duration(100)
			continue
		}

		// Get message from the queue
		message = &c.messages[0]
		c.messages = c.messages[1:]

		// Add messages to locked messages list
		messageReceived = true
		c.Lock.Unlock()
	}

	return message, nil
}

// Send method are sends a message into the queue.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId string    (optional) transaction id to trace execution through call chain.
//	- envelope *cqueues.MessageEnvelope  a message envelop to be sent.
// Returns: error or nil for success.
func (c *MqttMessageQueue) Send(ctx context.Context, correlationId string, envelop *cqueues.MessageEnvelope) error {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return err
	}

	c.Counters.IncrementOne(ctx, "queue."+c.Name()+".sent_messages")
	c.Logger.Debug(ctx, envelop.CorrelationId, "Sent message %s via %s", envelop.String(), c.Name())

	msg, err := c.fromMessage(envelop)
	if err != nil {
		return err
	}

	topic := c.topic
	if topic == "" {
		topic = c.Name()
	}

	err = c.Connection.Publish(ctx, topic, c.qos, c.retain, msg)
	if err != nil {
		c.Logger.Error(ctx, envelop.CorrelationId, err, "Failed to send message via %s", c.Name())
		return err
	}

	return nil
}

// RenewLock method are renews a lock on a message that makes it invisible from other receivers in the queue.
// This method is usually used to extend the message processing time.
// Important: This method is not supported by MQTT.
// Parameters:
//	- ctx context.Context	operation context.
//	- message   *cqueues.MessageEnvelope    a message to extend its lock.
//	- lockTimeout  time.Duration  a locking timeout in milliseconds.
// Returns: error
// receives an error or nil for success.
func (c *MqttMessageQueue) RenewLock(ctx context.Context, message *cqueues.MessageEnvelope, lockTimeout time.Duration) (err error) {
	// Not supported
	return nil
}

// Complete method are permanently removes a message from the queue.
// This method is usually used to remove the message after successful processing.
// Important: This method is not supported by MQTT.
// Parameters:
//	- ctx context.Context	operation context.
//	- message  *cqueues.MessageEnvelope a message to remove.
// Returns: error
// error or nil for success.
func (c *MqttMessageQueue) Complete(ctx context.Context, message *cqueues.MessageEnvelope) error {
	// Not supported
	return nil
}

// Abandon method are returnes message into the queue and makes it available for all subscribers to receive it again.
// This method is usually used to return a message which could not be processed at the moment
// to repeat the attempt. Messages that cause unrecoverable errors shall be removed permanently
// or/and send to dead letter queue.
// Important: This method is not supported by MQTT.
// Parameters:
//	- ctx context.Context	operation context.
//	- message *cqueues.MessageEnvelope  a message to return.
// Returns: error
//  error or nil for success.
func (c *MqttMessageQueue) Abandon(ctx context.Context, message *cqueues.MessageEnvelope) error {
	// Not supported
	return nil
}

// Permanently removes a message from the queue and sends it to dead letter queue.
// Important: This method is not supported by MQTT.
// Parameters:
//	- ctx context.Context	operation context.
//	- message  *cqueues.MessageEnvelope a message to be removed.
// Returns: error
//  error or nil for success.
func (c *MqttMessageQueue) MoveToDeadLetter(ctx context.Context, message *cqueues.MessageEnvelope) error {
	// Not supported
	return nil
}

func (c *MqttMessageQueue) sendMessageToReceiver(ctx context.Context, receiver cqueues.IMessageReceiver, message *cqueues.MessageEnvelope) {
	correlationId := message.CorrelationId

	defer func() {
		if r := recover(); r != nil {
			err := fmt.Sprintf("%v", r)
			c.Logger.Error(ctx, correlationId, nil, "Failed to process the message - "+err)
		}
	}()

	err := receiver.ReceiveMessage(ctx, message, c)
	if err != nil {
		c.Logger.Error(ctx, correlationId, err, "Failed to process the message")
	}
}

// Listens for incoming messages and blocks the current thread until queue is closed.
// Parameters:
//	- ctx context.Context	operation context.
//  - correlationId   string  (optional) transaction id to trace execution through call chain.
//  - receiver    cqueues.IMessageReceiver      a receiver to receive incoming messages.
//
// See IMessageReceiver
// See receive
func (c *MqttMessageQueue) Listen(ctx context.Context, correlationId string, receiver cqueues.IMessageReceiver) error {
	err := c.CheckOpen(correlationId)
	if err != nil {
		return err
	}

	// Subscribe if needed
	err = c.subscribe(ctx, correlationId)
	if err != nil {
		return err
	}

	c.Logger.Trace(ctx, "", "Started listening messages at %s", c.Name())

	// Get all collected messages
	c.Lock.Lock()
	batchMessages := c.messages
	c.messages = []cqueues.MessageEnvelope{}
	c.Lock.Unlock()

	// Resend collected messages to receiver
	for _, message := range batchMessages {
		receiver.ReceiveMessage(ctx, &message, c)
	}

	// Set the receiver
	c.Lock.Lock()
	c.receiver = receiver
	c.Lock.Unlock()

	return nil
}

// EndListen method are ends listening for incoming messages.
// When this method is call listen unblocks the thread and execution continues.
// Parameters:
//	- ctx context.Context	operation context.
//	- correlationId  string   (optional) transaction id to trace execution through call chain.
func (c *MqttMessageQueue) EndListen(ctx context.Context, correlationId string) {
	c.Lock.Lock()
	c.receiver = nil
	c.Lock.Unlock()
}
